mypy/plugins/dataclasses.py

```diff
@@ -0,0 +1,279 @@
+            'frozen': _get_decorator_bool_argument(self._ctx, 'frozen', False),
+        }
+
+        if decorator_arguments['init']:
+            _add_method(
+                ctx,
+                '__init__',
+                args=[attr.to_argument(info) for attr in attributes if attr.is_in_init],
+                return_type=NoneTyp(),
+            )
+            for stmt in self._ctx.cls.defs.body:
+                if isinstance(stmt, Decorator) and stmt.func.is_class:
+                    func_type = stmt.func.type
+                    if isinstance(func_type, CallableType):
+                        func_type.arg_types[0] = self._ctx.api.class_type(self._ctx.cls.info)
+
+        # Add an eq method, but only if the class doesn't already have one.
+        if decorator_arguments['eq'] and info.get('__eq__') is None:
+            cmp_tvar_def = TypeVarDef('T', 'T', 1, [], ctx.api.named_type('__builtins__.object'))
+            cmp_other_type = TypeVarType(cmp_tvar_def)
+            cmp_return_type = ctx.api.named_type('__builtins__.bool')
+
+            for method_name in ['__eq__', '__ne__']:
+                _add_method(
+                    ctx,
+                    method_name,
+                    args=[Argument(Var('other', cmp_other_type), cmp_other_type, None, ARG_POS)],
+                    return_type=cmp_return_type,
+                    self_type=cmp_other_type,
+                    tvar_def=cmp_tvar_def,
+                )
+
+        # Add <,>,<=,>=, but only if the class has an eq method.
+        if decorator_arguments['order']:
+            if not decorator_arguments['eq']:
+                ctx.api.fail('eq must be True if order is True', ctx.cls)
+
+            order_tvar_def = TypeVarDef('T', 'T', 1, [], ctx.api.named_type('__builtins__.object'))
+            order_other_type = TypeVarType(order_tvar_def)
+            order_return_type = ctx.api.named_type('__builtins__.bool')
+            order_args = [
+                Argument(Var('other', order_other_type), order_other_type, None, ARG_POS)
+            ]
+
+            for method_name in ['__lt__', '__gt__', '__le__', '__ge__']:
+                existing_method = info.get(method_name)
+                if existing_method is not None:
+                    assert existing_method.node
+                    ctx.api.fail(
+                        'You may not have a custom %s method when order=True' % method_name,
+                        existing_method.node,
+                    )
+
+                _add_method(
+                    ctx,
+                    method_name,
+                    args=order_args,
+                    return_type=order_return_type,
+                    self_type=order_other_type,
+                    tvar_def=order_tvar_def,
+                )
+
+        if decorator_arguments['frozen']:
+            self._freeze(attributes)
+
+        info.metadata['dataclass'] = {
+            'attributes': {attr.name: attr.serialize() for attr in attributes},
```

Comment:
This probably should be an OrderedDict in order to preserve order in pre-3.6 Python versions. I think that's what's causing the test failure on 3.4 and 3.5 in Travis.

Reply:
Ahh, that's right! I'll make the change.

Reply:
Change made. :)

Reply:
Thanks! For the future though, it's generally easier on the reviewer to push new commits instead of amending your previous commit and force-pushing as you're doing now. That way, it's easier for the reviewer to see what changes were made in response to the review.

We squash the commits anyway before merging, so it's OK to have a lot of commits in your PR branch.